### PR TITLE
feat(sink): support jdbc sink async truncate

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

After https://github.com/risingwavelabs/risingwave/pull/13137

The return type of the `write` method of `SinkWriter` java interface is changed to `bool`, where `true` indicates that the batch is persisted in the external sink and is safe to truncate the log reader.

A new item `BatchWritten` is added to the sink writer response, to indicate that the batch with the batch id has been persisted. The remote sink is refactored to poll the response and truncate asynchronously.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/sink/remote.rs [hunk 1]
```diff
@@ -64,28 +67,54 @@ use crate::ConnectorParams;
 macro_rules! def_remote_sink {
     () => {
         def_remote_sink! {
-            { ElasticSearch, ElasticSearchSink, "elasticsearch" },
-            { Cassandra, CassandraSink, "cassandra" },
-            { Jdbc, JdbcSink, "jdbc" },
+            { ElasticSearch, ElasticSearchSink, "elasticsearch" }
+            { Cassandra, CassandraSink, "cassandra" }
+            { Jdbc, JdbcSink, "jdbc", |desc| {
+                desc.sink_type.is_append_only()
```
Comment (hzxa21) [inline src/connector/src/sink/remote.rs:73]: Are we planning to enable upsert jdbc sink decoupling as well?

File: src/connector/src/sink/remote.rs [hunk 2]
```diff
@@ -241,80 +274,162 @@ impl LogSinker for RemoteLogSinker {
...
+                                response:
+                                    Some(sink_writer_stream_response::Response::Commit(
+                                        sink_writer_stream_response::CommitResponse {
+                                            epoch,
+                                            metadata,
+                                        },
+                                    )),
+                            } => {
+                                if let Some(metadata) = metadata {
+                                    warn!("get unexpected non-empty metadata: {:?}", metadata);
```
Comment (hzxa21) [inline src/connector/src/sink/remote.rs:360]: Is this really unexpected? I saw that java iceberg sink will return non-empty metadata.

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

File: src/connector/src/sink/remote.rs [hunk 3]
```diff
@@ -241,80 +274,162 @@ impl LogSinker for RemoteLogSinker {
...
 
+        let poll_consume_log_and_sink = async move {
             log_reader.init().await?;
 
-            loop {
-                let (epoch, item): (u64, LogStoreReadItem) =
-                    log_reader.next_item().map_err(SinkError::Internal).await?;
+            async fn truncate_matched_offset(
+                queue: &mut VecDeque<(TruncateOffset, Option<Instant>)>,
+                offset: TruncateOffset,
```
Comment (StrikeW) [inline src/connector/src/sink/remote.rs:284]: may rename to `persisted_offset` or `committed_offset`?

**Review by StrikeW — APPROVED** [review]
Rest lgtm. I think the related issue is https://github.com/risingwavelabs/risingwave/issues/12060. 
Btw I have a quick question about the failure isolation, what is the expected behavior if a sink fails to commit? For the case of JDBC sink fails due to incorrect data, I think we can avoid triggering a recovery with the help of log store. But does the sink will retry its work continuously?

File: proto/connector_service.proto [hunk 4]
```diff
@@ -88,9 +88,15 @@ message SinkWriterStreamResponse {
     SinkMetadata metadata = 2;
   }
 
+  message BatchWrittenResponse {
+    uint64 epoch = 1;
+    uint64 batch_id = 2;
```
[the remark this answers is not in the document]
Comment (wenym1) [reply proto/connector_service.proto:93]: `batch_id` is set as the `chunk_id` received from `LogReader`. Inside the `LogReader` we ensure that the chunk id is monotonically increasing.

File: src/connector/src/sink/remote.rs (hunk 2 above)
Comment (wenym1) [reply src/connector/src/sink/remote.rs:362]: iceberg sink uses `CoordinatedRemoteSinkWriter` and will not be affected by this PR. Sink that use this `RemoteLogSinker` assume that they are not coordinated and should not return non-empty metadata.

File: src/connector/src/sink/remote.rs (hunk 1 above)
Comment (wenym1) [reply src/connector/src/sink/remote.rs:73]: May keep the same as other sinks like kafka, kinesis for now. Current implementation does not support rewind on failure yet, and enabling it by default does not help resolve much of the current problems. 

I think we can enable sink decouple for these sinks by default after rewind on failure is supported.

**Pull request merged**